File: Cred.R
#install.packages("DMwR")
#install.packages("caret")
#install.packages("unbalanced")
#install.packages("c50")
#install.packages("dummies")
#install.packages("gridExtra")
#install.packages("psych")


setwd("F:/proj1")
#library list
libs=c("ggplot2","corrgram","DMwR","caret","randomForest","unbalanced","c50","dummies","MASS","gridExtra","psych")

#load Libraries
lapply(libs,require,character.only=TRUE)

#read data
raw_dat=read.csv("credit-card-data.csv",header = T,na.strings = c(" ","","NA"))       

#missing value analysis
miss_val=data.frame(apply(raw_dat,2,function(x){sum(is.na(x))}))

miss_val$columns=row.names(miss_val)
row.names(miss_val)=NULL
names(miss_val)[1]="percentage"
miss_val$percentage = (miss_val$percentage/nrow(raw_dat))*100
miss_val$percentage =miss_val[order(-miss_val$percentage),]


#1 - 139.50979
#9 - 311.96341
#mean =	864.35442
#raw_dat$MINIMUM_PAYMENTS[is.na(raw_dat$MINIMUM_PAYMENTS)]=mean(raw_dat$MINIMUM_PAYMENTS,na.rm = T)

#median = 312.56064
#impute Missing values with median
raw_dat$MINIMUM_PAYMENTS[is.na(raw_dat$MINIMUM_PAYMENTS)]=median(raw_dat$MINIMUM_PAYMENTS,na.rm = T)


#outlier analysis

indx_val = sapply(raw_dat, is.numeric)
num_data = raw_dat[,indx_val]

cols= colnames(num_data)
#plot ggplot
for (i in 1:length(cols)) {
  assign(paste0("plot",i),ggplot(aes_string(y=(cols[i]),x="CUST_ID"),dat=subset(raw_dat))+
           stat_boxplot(geom='errorbar',width=0.5)+
  geom_boxplot(outlier.colour="red",fill="blue",outlier.shape=18,outlier.size=1,notch=F)+
    theme(legend.position="bottom")+
    labs(y=cols[i],x="CUST_ID"))
}

gridExtra::grid.arrange(plot1,plot2,plot3,ncol=3)



#remove outliers
for (i in cols) {
  
  val=raw_dat[,i][raw_dat[,i]%in%boxplot.stats(raw_dat[,i])$out]
  raw_dat=raw_dat[which(!raw_dat[,i]%in%val),]
}

#derive an enriched data frame as per KPI's

new_data=data.frame("CUST_ID"=raw_dat$CUST_ID,"monthly_average_purchase"= raw_dat$PURCHASES,
                        "CASH_ADVANCE"=raw_dat$CASH_ADVANCE,"ONEOFF_PURCHASES"=raw_dat$ONEOFF_PURCHASES,
                        "INSTALLMENTS_PURCHASES"=raw_dat$INSTALLMENTS_PURCHASES,
                        "average_amount_per_purchase"=raw_dat$PURCHASES_TRX,"CASH_ADVANCE_TRX"=raw_dat$CASH_ADVANCE_TRX,
                        "LIMIT_usage"=raw_dat$BALANCE/raw_dat$CREDIT_LIMIT,
                        "payments_to_minimum_payments_ratio"= raw_dat$PAYMENTS/raw_dat$MINIMUM_PAYMENTS )
#get only the numeric variables
indx_val = sapply(new_data, is.numeric)
num_data = new_data[,indx_val]



#correlation
corrgram(new_data[,indx_val],order = F,upper.panel = panel.shade,text.panel = panel.text,main="correlation")


#removing variable "CUST_ID" as it is in factor type and not usefull for building a model.
new_data=subset(new_data,select=-c(CUST_ID))

cols=colnames(new_data)

#normalization
for (i in cols) {
  new_data[,i]=(new_data[,i]-min(new_data[,i]))/(max(new_data[,i])-min(new_data[,i]))
}


#standardization
for (i in cols) {
  new_data[,i]=(new_data[,i]-mean(new_data[,i]))/sd(new_data[,i])
}
rmExcept("new_data")

#FACTOR ANALYSIS

# find principle component to get the no. of factors to select

pc=princomp(new_data)

summary(pc)

#plot a bar graph to conform the number the factors
plot(pc)



fa=factanal(new_data,factors = 4 )

print(fa,cutoff=.3,sort=T)

#temp=data.frame(scale(new_data[-5]))

#Kmeans Clustering

library(NbClust)
km=kmeans(new_data,4,nstart = 25)

acc=table(temp$average_amount_per_purchase,km$cluster)

File: Cred.py
import numpy as np
import pandas as pd
import os
import matplotlib.pyplot as plt
import seaborn as sns
import random
from factor_analyzer import FactorAnalyzer
from sklearn.decomposition import *
from sklearn.cluster import KMeans

os.chdir("F:/proj1")

#read data
raw_dat=pd.read_csv("credit-card-data.csv")

#find missing value and its percentage
miss_val=pd.DataFrame(raw_dat.isnull().sum())
miss_val=miss_val.reset_index()
miss_val=miss_val.rename(columns={'index': 'Variables',0:'missing_percentage'})
miss_val['missing_percentage']=(miss_val['missing_percentage']/len(raw_dat))*100

miss_val=miss_val.rename(columns={'index': 'Variables',0:'missing_percentage'})

#collection of numeric columns
c=['BALANCE','BALANCE_FREQUENCY','PURCHASES','ONEOFF_PURCHASES','INSTALLMENTS_PURCHASES','CASH_ADVANCE','PURCHASES_FREQUENCY',
   'ONEOFF_PURCHASES_FREQUENCY','PURCHASES_INSTALLMENTS_FREQUENCY','CASH_ADVANCE_FREQUENCY','CASH_ADVANCE_TRX','PURCHASES_TRX',
   'CREDIT_LIMIT','PAYMENTS','MINIMUM_PAYMENTS','PRC_FULL_PAYMENT','TENURE']


#finding and removing outliers
for i in c:
    q75,q25=np.percentile(raw_dat.loc[:,i],[75,25])
    iqr=q75-q25
    min=q25-(iqr*1.5)
    max=q75+(iqr*1.5)
    print(min)
    print(max)
    raw_dat=raw_dat.drop(raw_dat[raw_dat.loc[:,i]<min].index)
    raw_dat=raw_dat.drop(raw_dat[raw_dat.loc[:,i]>max].index)
    raw_dat=raw_dat.reset_index(drop=True)


f,ax=plt.subplots(figsize=(7,5))
corr=raw_dat.corr()
sns.heatmap(corr,mask=np.zeros_like(corr,dtype=np.bool),cmap=sns.diverging_palette(220,10,as_cmap=True),square=True,ax=ax)


#derving data from the original data according to KPI
derived_data={'CUST_ID':raw_dat['CUST_ID'],'monthly_average_purchase':raw_dat['PURCHASES'],'CASH_ADVANCE':raw_dat['CASH_ADVANCE'],
             'ONEOFF_PURCHASES':raw_dat['ONEOFF_PURCHASES'],'INSTALLMENTS_PURCHASES':raw_dat['INSTALLMENTS_PURCHASES'],
              'PURCHASES_TRX':raw_dat['PURCHASES_TRX'],'CASH_ADVANCE_TRX':raw_dat['CASH_ADVANCE_TRX'],
              'LIMIT_USAGE':raw_dat['BALANCE']/raw_dat['CREDIT_LIMIT'],
              'PAYMENTS_TO_MIN_PAYMENTS_RATIO':raw_dat['PAYMENTS']/raw_dat['MINIMUM_PAYMENTS']}

derived_data=pd.DataFrame(derived_data)

#plot boxplot to visualize outliers
sns.boxplot(derived_data['monthly_average_purchase'])
plt.boxplot(derived_data['CASH_ADVANCE'])
plt.boxplot(derived_data['ONEOFF_PURCHASES'])
plt.boxplot(derived_data['INSTALLMENTS_PURCHASES'])
plt.boxplot(derived_data['PURCHASES_TRX'])
plt.boxplot(derived_data['CASH_ADVANCE_TRX'])
plt.boxplot(derived_data['LIMIT_USAGE'])
sns.boxplot(derived_data['PAYMENTS_TO_MIN_PAYMENTS_RATIO'])

#removing variable "CUST_ID" as it is in factor type and not usefull for building a model.
derived_data=derived_data.drop(columns=['CUST_ID'])


#PLotting correlation map
f,ax=plt.subplots(figsize=(7,5))
corr=derived_data.corr()
sns.heatmap(corr,mask=np.zeros_like(corr,dtype=np.bool),cmap=sns.diverging_palette(220,10,as_cmap=True),square=True,ax=ax)

#normality check
plt.hist(derived_data['PURCHASES_TRX'],bins='auto')

#normalization
for i in cols:
    print(i)
    derived_data[i]=(derived_data[i]-derived_data[i].min())/(derived_data[i].max()-derived_data[i].min())

#standarization
for i in cols:
    print(i)
    derived_data[i]=(derived_data[i]-derived_data[i].mean())/(derived_data[i].std())


#factor analysis for getting ideal number factors to select
fa = FactorAnalyzer(rotation=None)
fa.fit(derived_data)
ev ,v= fa.get_eigenvalues()

#eigen values
ev


#plot a elbow graph for getting number factors
plt.plot(range(1,derived_data.shape[1]+1),ev)
plt.title('Scree Plot')
plt.xlabel('Factors')
plt.ylabel('Eigenvalue')
plt.grid()
plt.show()	


#factor analysis with number factors =4
fa = FactorAnalyzer(n_factors=2,rotation="varimax")
fa.fit(derived_data)


fa.get_factor_variance()
#cummuilative variance=47%

#loadings per each factor
fa.loadings_

#Kmeans clustering


clusters=KMeans(init="random",n_clusters=2)
clusters.fit(derived_data)
